### PR TITLE
Add parallelio

### PR DESCRIPTION
**Pull request by xylar**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by xylar** [comment]
This appears to be needed for https://github.com/conda-forge/esmf-feedstock/pull/78.  Without an external `parallelio`, ESMF attempts to vendor its own internal version (which is not building successfully).

File: recipes/parallelio/conda_build_config.yaml
Comment (xylar) [inline recipes/parallelio/conda_build_config.yaml:11]: As commented, here, I wasn't able to get `conda mambabuild` to run locally without these pins, taken from `conda-forge-pinning`.  I kept seeing errors about not finding:
```
netcdf-fortran mpi_openmpi.*
```
I could not figure out why it was getting confused about a constraint on a build string that it seems to think is on the version.

Comment (xylar) [reply recipes/parallelio/conda_build_config.yaml:11]: I will remove these as soon as this is in its own feedstock.

**Comment by xylar** [comment]
@edwardhartnett and @jedwards4b, I'm hoping to add a recipe for ParallelIO on conda-forge.  I have a lot of experience maintaining recipes here but I'm better with python packages than c/c++/fortran libraries.  Still, I'm a developer with E3SM and can get help from that team if need be.

I'm wondering if you have input on this recipe, particularly the build script (`build.sh`).  You would both be welcome to become maintainers of the package but I'm assuming that's not something you're interested in doing.

File: recipes/parallelio/build.sh
Comment (xylar) [inline recipes/parallelio/build.sh:17]: For now, I am only including NetCDF and not PNetCDF because there isn't a conda-forge package for PNetCDF.  I hope to follow up soon with a PNetCDF package at which point I will make a build of ParallelIO that includes PNetCDF support.

File: recipes/parallelio/build.sh
Comment (xylar) [inline recipes/parallelio/build.sh:22]: I had trouble with some tests when I ran locally:
```
[ 71%] Linking C executable piodecomptest
/home/xylar/miniconda3/conda-bld/parallelio_1661760167351/_build_env/bin/../lib/gcc/x86_64-conda-linux-gnu/12.1.0/../../../../x86_64-conda-linux-gnu/bin/ld: CMakeFiles/piodecomptest.dir/mpi_argp.c.o: in function `argp_usage':
mpi_argp.c:(.text.argp_usage+0x0): multiple definition of `argp_usage'; CMakeFiles/piodecomptest.dir/piodecomptest.c.o:piodecomptest.c:(.text.argp_usage+0x0): first defined here
/home/xylar/miniconda3/conda-bld/parallelio_1661760167351/_build_env/bin/../lib/gcc/x86_64-conda-linux-gnu/12.1.0/../../../../x86_64-conda-linux-gnu/bin/ld: CMakeFiles/piodecomptest.dir/mpi_argp.c.o: in function `_option_is_short':
mpi_argp.c:(.text._option_is_short+0x0): multiple definition of `_option_is_short'; CMakeFiles/piodecomptest.dir/piodecomptest.c.o:piodecomptest.c:(.text._option_is_short+0x0): first defined here
/home/xylar/miniconda3/conda-bld/parallelio_1661760167351/_build_env/bin/../lib/gcc/x86_64-conda-linux-gnu/12.1.0/../../../../x86_64-conda-linux-gnu/bin/ld: CMakeFiles/piodecomptest.dir/mpi_argp.c.o: in function `_option_is_end':
mpi_argp.c:(.text._option_is_end+0x0): multiple definition of `_option_is_end'; CMakeFiles/piodecomptest.dir/piodecomptest.c.o:piodecomptest.c:(.text._option_is_end+0x0): first defined here
```
I don't yet feel like I have the expertise to debug this but I hope to figure out whether I have something misconfigured here (most likely a missing cmake flag) or if there is a problem with the test.

Comment (jedwards4b) [reply recipes/parallelio/build.sh:22]: I've not seen this error before, but the problem appears to be in <argp.h>

**Comment by jedwards4b** [comment]
Yes you can list me as a maintainer - my github id is jedwards4b.
   
I think that it is worth mentioning that we have a [spack package](https://github.com/spack/spack/tree/develop/var/spack/repos/builtin/packages/parallelio) for parallelio and e3sm has the [scorpio fork](https://github.com/E3SM-Project/scorpio) of the repository.

**Comment by xylar** [comment]
Thanks @jedwards4b, I've added you.  Your insights as a fellow maintainer will be really valuable!

> I think that it is worth mentioning that we have a [spack package](https://github.com/spack/spack/tree/develop/var/spack/repos/builtin/packages/parallelio) for parallelio and e3sm has the [scorpio fork](https://github.com/E3SM-Project/scorpio) of the repository.

Yes, I'm aware of ParallelIO being on Spack as well.  On HPC systems in most contexts, that's likely to be the preferred way to build.  In cases where folks want a conda environment with a bunch of python packages for analysis, including ESMPy, conda will continue to be an appealing alternative.  The latest ESMF seems like it will be happier with an external PIO.

File: recipes/parallelio/build.sh
Comment (xylar) [reply recipes/parallelio/build.sh:24]: I have commented out that test and can get the rest to build but most (all?) MPI tests fail on run because I don't have enough cores on my laptop.  I think this is going to be even worse for CI.  So I think we should let `ctest` be for now.

**Comment by xylar** [comment]
@conda-forge/help-c-cpp, this is ready for review.  Your recommendations to improve the recipe are most welcome.  My expertise is more with python than with c/c++ packages.  But I use the ESMF package frequently, and that package would very much benefit from this one.

File: recipes/parallelio/conda_build_config.yaml
Comment (carterbox) [reply recipes/parallelio/conda_build_config.yaml:11]:
```
netcdf-fortran mpi_openmpi.*
```
Above is an invalid package spec. It should be something like this:

```
netcdf-fortran * mpi_openmpi_*
```
or
```
netcdf-fortran =*=mpi_openmpi_*
```

Comment (carterbox) [reply recipes/parallelio/conda_build_config.yaml:11]: I assume this problem is because you weren't using the `build-locally.py` script? This script does environment/recipe augmentation stuff in addition to starting a docker container.

File: recipes/parallelio/meta.yaml
Comment (carterbox) [inline recipes/parallelio/meta.yaml:48]: Check the logs. The link inspector thinks you are overdepending on netcdf-fortran and hdf5 because it can't find any dynamic links to these libraries. Are y'all doing static linking?

**Review by carterbox — CHANGES_REQUESTED** [review]

File: recipes/parallelio/meta.yaml
Comment (carterbox) [inline recipes/parallelio/meta.yaml:39]: The openmp packages are also unused according to the link inspector.

File: recipes/parallelio/conda_build_config.yaml
Comment (xylar) [reply recipes/parallelio/conda_build_config.yaml]: I can't install docker on my laptop because it interferes with my ability to use wifi on my laptop at work.  I can take these out and see if CI does okay.

Comment (xylar) [reply recipes/parallelio/conda_build_config.yaml]: I see you went ahead and took these out, and things built.  Thanks.  Thanks for the tip about `build-locally.py`. I'll see if I can find a solution where I can install Docker and still use LANL's wifi.

File: recipes/parallelio/meta.yaml
Comment (xylar) [reply recipes/parallelio/meta.yaml:39]: Okay, I'm fine with removing these.  It may be that I am not enabling OpenMP support properly but this can be something we add down the road.

File: recipes/parallelio/meta.yaml
Comment (xylar) [reply recipes/parallelio/meta.yaml:48]: @carterbox, you may be right that `hdf5` isn't needed.  It isn't explicitly referenced in the `cmake` command.

Regarding `netcdf-fortran`, this is clearly a dependency.  The problem could easily lie with that library rather than here.  It would appear the static and shared libraries are both being included in that package:
https://github.com/conda-forge/netcdf-fortran-feedstock/blob/main/recipe/build.sh#L32-L58
As a result, presumably the static library is being linked here instead of the dynamic?

I won't have time for awhile but I can look into splitting that package into only the dynamic library and a separate `netcdf-fortran-static` package.

**Comment by xylar** [comment]
@carterbox, I really appreciate your careful review and the changes you pushed.  I'm going to try to fix `netcdf-fortran` and then I will remove the other dependencies you flagged as unneeded and ask for a re-review.

File: recipes/parallelio/conda_build_config.yaml
Comment (carterbox) [reply recipes/parallelio/conda_build_config.yaml]: Pretty it works with podman too in case that helps.

Comment (xylar) [reply recipes/parallelio/conda_build_config.yaml]: That could be worth a try, thanks!

File: recipes/parallelio/meta.yaml
Comment (xylar) [reply recipes/parallelio/meta.yaml:48]: This is turning out to be more complicated than I hoped.  The c library has not yet been split into separate dynamic and static packages:
https://github.com/conda-forge/libnetcdf-feedstock/pull/135
That's presumably a prerequisite for splitting the fortran library, since it has the c library as a dependency.

**Comment by xylar** [comment]
@carterbox, I could use some help here.  I believe my fellow maintainers and I have made the changes needed so that the latest `libnetcdf` and `netcdf-fortran` package no longer include static libraries.

I have removed unneeded HDF5 and OpenMP dependencies.

Now, we are seeing the same weird glitch here that I was seeing in local (non-docker) builds before:
```
 - libgfortran5 >=11.3.0
 - netcdf-fortran mpi_openmpi_.*
 - libgfortran 5.*
 - libnetcdf >=4.8.1,<4.8.2.0a0 mpi_openmpi_*
 - openmpi >=4.1.4,<5.0a0

with channels:
 - local
 - conda-forge
 - conda-forge

The reported errors are:
- Encountered problems while solving:
-   - nothing provides requested netcdf-fortran mpi_openmpi_.*

```
`mamba` is seeing the build constraint as a version constraint.  Any clue?

**Comment by carterbox** [comment]
I think the problem is not with this recipe, but with a broken build of netcdf-fortran that had an incorrect run_export which caused downstream packages to be broken.

**Comment by xylar** [comment]
@carterbox, thanks, I'll rebuild that release of netcdf-fortran to fix it.  I think you're likely right.

**Comment by carterbox** [comment]
No. You should wait.

**Comment by xylar** [comment]
> No. You should wait.

Can you be more specific?

**Comment by carterbox** [comment]
Please ping the help team again when the repodata has been patched and the build environment is solvable. Thanks!

**Comment by xylar** [comment]
@conda-forge-admin, please restart ci

**Comment by xylar** [comment]
@conda-forge/help-c-cpp, this is ready for another look when you have time.

it seems that the issues with `netcdf-fortran` have been fixed by a new build.  I have also requested that the builds with bad `run_export` get marked as broken, see https://github.com/conda-forge/admin-requests/pull/497.  We can wait for that PR to get merged if necessary.

**Comment by xylar** [comment]
@conda-forge/help-c-cpp and @carterbox, I'm keen to have this re-reviewed when you can.

File: recipes/parallelio/meta.yaml
Comment (carterbox) [reply recipes/parallelio/meta.yaml]: Looks like the static libs were removed after the latest build on this PR. I will restart the build to see if it builds successfully without static libs for netcdf-fortran.

https://github.com/conda-forge/netcdf-fortran-feedstock/pull/76

**Comment by carterbox** [comment]
Restarting build to check compatibility with latest netcdf builds.

File: recipes/parallelio/meta.yaml
Comment (carterbox) [reply recipes/parallelio/meta.yaml]: [Logs are still showing no dynamic linkage to netcdf-fortran.
](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=598878&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823&l=1486) We should be able to detect a link to `libnetcdff.so.7` if it is actually being used.

**Comment by carterbox** [comment]
I'm stumped about the non-detectable links / overdepending warning for netcdf-fortran. I've added this library to the `build/ignore_run_exports_from` section. Since it looks like we're building against the version with static libs removed, it seems to me that maybe only the headers are used(?) or the build system for this library requests packages that aren't actually needed.

**Pull request merged**
